File: hibernate-query-language/src/main/java/io/github/robertomike/hefesto/constructors/ConstructJoinFetch.java
package io.github.robertomike.hefesto.constructors;

import io.github.robertomike.hefesto.actions.JoinFetch;
import io.github.robertomike.hefesto.builders.Hefesto;

public class ConstructJoinFetch extends Construct<JoinFetch> {
    public String construct(Hefesto<?> builder) {
        StringBuilder joinQuery = new StringBuilder();

        items.forEach((value) -> {
            if (!joinQuery.isEmpty()) {
                joinQuery.append(" ");
            }

            joinQuery.append(apply(builder, value));
        });

        return joinQuery.toString();
    }

    public String apply(Hefesto<?> builder, JoinFetch join) {
        return join.getJoinType() + " join fetch " +
                (join.isNested() ? "" : builder.getAcronymTable() + ".") + join.getRelationship() +
                (join.getAlias() != null ? " as " + join.getAlias() : "");
    }
}

File: hibernate-query-language/src/main/kotlin/io/github/robertomike/hefesto/constructors/ConstructJoinFetch.kt
package io.github.robertomike.hefesto.constructors

import io.github.robertomike.hefesto.actions.JoinFetch
import io.github.robertomike.hefesto.builders.Hefesto

class ConstructJoinFetch : Construct<JoinFetch>() {
    fun construct(builder: Hefesto<*>): String {
        val joinQuery = StringBuilder()

        items.forEach { value ->
            if (joinQuery.isNotEmpty()) {
                joinQuery.append(" ")
            }

            joinQuery.append(apply(builder, value))
        }

        return joinQuery.toString()
    }

    fun apply(builder: Hefesto<*>, join: JoinFetch): String {
        return "${join.joinType} join fetch ${if (join.nested) "" else "${builder.acronymTable}."}${join.relationship}${if (join.alias != null) " as ${join.alias}" else ""}"
    }
}
